pkg/controller/volume/persistentvolume/pv_controller.go

```diff
@@ -314,8 +314,18 @@ func (ctrl *PersistentVolumeController) syncUnboundClaim(claim *v1.PersistentVol
 			return err
 		}
 
+		var node *v1.Node
+		if selectedNode, ok := claim.Annotations[pvutil.AnnSelectedNode]; ok {
+			node, err = ctrl.NodeLister.Get(selectedNode)
+			if err != nil {
+				strerr := fmt.Sprintf("Failed to get selected node: %v", err)
+				klog.V(3).Infof("unexpected error getting selected node %q for claim %q: %v", selectedNode, claimToClaimKey(claim), err)
+				ctrl.eventRecorder.Event(claim, v1.EventTypeWarning, events.FailedBinding, strerr)
+				return err
+			}
+		}
 		// [Unit test set 1]
-		volume, err := ctrl.volumes.findBestMatchForClaim(claim, delayBinding)
+		volume, err := ctrl.volumes.findBestMatchForClaim(claim, node, delayBinding)
```

Comment:
Thank you for the PR!

If selectedNode is set, then that means we want dynamic provisioning.  So in all cases, I think we should return `volume = nil` from this call.  I think it would be cleaner if we added the annotation check inside `findBestMatchForClaim` instead of here. cc @cofyc

Reply:
Hm actually the issue is a little trickier than I thought. Reading your original issue, I guess the problem is if PV doesn't exist by the time PV controller checks, then we will set selected-node annotation.  But if your volume plugin doesn't actually support provisioning, then it's going to get stuck waiting for dynamic provisioning forever with the change I suggested.

I need to think more about this case, but I think setting "kubernetes.io/no-provisioner" in the StorageClass should prevent this.

Reply:
Thanks for your reply!

For the first part, do you mean that we should never choose a pv for those pvcs we want dynamic provisioning, and left it to the dynamic provision controller to set the provisioned volume name to the claim?
then, the second part is also fit, i think.

Reply:
The sequence of volume binding for a PVC is:
1. First check if any existing PVs exist that can match the PVC.
2. If no PVs match, then try dynamic provisioning.

So first question I have is do you have a dynamic provisioner?

If no, then you should set `provisioner: kubernetes.io/no-provisioner` so that we don't try dynamic provisioning and get stuck there forever.

If yes, then you have the limitation that:
* the PV should be created first or
* you prebind the PVC to the PV or
* use a different storageclass with no-provisioner set.

Either way, I think there is a bug where if selected-node is already set on a PVC, the scheduler should not try to find an existing PV and bind it.

Reply:
yeah, let us limit the problem to the scope that dynamic provisioner is present and no PV found when scheduling, so  
> selected-node is already set on a PVC

then another approach may be cleaner:
pv_controller should do no try to find a PV to bind when sync for these kind of PVCs, and the provisioner is **required** to pre-bind the PV created next for the PVC (as the csi-provisioner do).
I'll push another commit if you think so, :)

though CSI provide the normal provisioner for external, 
may this approach has potential compatibility issue that older dynamic provisioner is not strictly **required** to do that?

Reply:
Yes, I think the pv controller should not try to find a PV to bind to if selected-node is set.

I'm unsure about requiring the provisioner to prebind PVs that were created outside of provisioner.  In this case, it sounds like for this particular PVC, you do not want dynamic provisioning and you want a specific PV. In this case, I think you should set PVC.volumeName when you create the PVC.

Reply:
great, please help to check the new commit.